sklearn/ensemble/iforest.py

```diff
@@ -151,15 +151,28 @@ def fit(self, X, y=None, sample_weight=None):
         y = rnd.uniform(size=X.shape[0])
 
         # ensure that max_sample is in [1, n_samples]:
-        max_samples = self.max_samples
         n_samples = X.shape[0]
-        if max_samples > n_samples:
-            warn("max_samples (%s) is greater than the "
-                 "total number of samples (%s). max_samples "
-                 "will be set to n_samples for estimation."
-                 % (self.max_samples, n_samples))
-            max_samples = n_samples
-
+        max_samples = min(256, n_samples)
+
+        if self.max_samples != 'auto':
+            if isinstance(self.max_samples, (numbers.Integral, np.integer)):
+                if self.max_samples < 1:
+                    raise ValueError("max_features must be larger than zero.")
+
+                if self.max_samples > n_samples:
+                    warn("max_samples (%s) is greater than the "
+                         "total number of samples (%s). max_samples "
+                         "will be set to n_samples for estimation."
+                         % (self.max_samples, n_samples))
+                    max_samples = n_samples
+
+            else: # float
+                if 0. < self.max_samples <= 1.:
+                    max_samples = self.max_samples * n_samples
+                else:
+                    raise ValueError('max_samples must be in (0, 1]')
+
+        self.base_estimator.max_depth = int(np.ceil(np.log2(max(max_samples, 2))))
```

Comment:
I'm not sure if it is `self.base_estimator` or `self.base_estimator_` that you have to consider.